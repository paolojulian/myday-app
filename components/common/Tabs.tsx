import ThemedText from '@/components/common/ThemedText';
import ThemedView from '@/components/common/ThemedView';
import { colors } from '@/constants/Colors';
import { StyleSheet, TouchableWithoutFeedback } from 'react-native';

type TabsProps<T> = {
  onSelect: (item: T) => void;
  selectedItem: T;
  items: T[];
};

export default function Tabs<T extends string>({ onSelect, selectedItem, items }: TabsProps<T>) {
  return (
    <ThemedView style={styles.container}>
      {items.map(item => (
        <TouchableWithoutFeedback key={item} onPress={() => onSelect(item)}>
          <ThemedView
            style={[
              styles.item,
              {
                ...(selectedItem === item && styles.itemActive),
              },
            ]}
          >
            <ThemedText style={{ ...(selectedItem === item && styles.itemTextActive) }}>
              {item}
            </ThemedText>
          </ThemedView>
        </TouchableWithoutFeedback>
      ))}
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    backgroundColor: colors.whiteSmoke,
    borderRadius: 100,
  },
  item: {
    flex: 1,
    padding: 16,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 100,
    color: colors.black,
  },
  itemActive: {
    backgroundColor: colors.black,
  },
  itemTextActive: {
    color: colors.white,
  },
});
