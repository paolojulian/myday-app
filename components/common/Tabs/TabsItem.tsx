import { TabsVariant } from '@/components/common/Tabs/Tabs';
import ThemedText from '@/components/common/ThemedText';
import ThemedView from '@/components/common/ThemedView';
import { colors } from '@/constants/Colors';
import React from 'react';
import { StyleSheet, TouchableWithoutFeedback } from 'react-native';

export const TABS_ITEM_TEST_IDS = {
  containerBtn: (id: string, isSelected: boolean) => `tabs-item-container-btn-${id}-${isSelected}`,
  text: (id: string) => `tabs-item-text-${id}`,
};

export type TabItem<T extends string | number> = {
  key: T;
  value: string;
};

type TabsItemProps<T extends string | number> = {
  item: TabItem<T>;
  onSelect: (key: T) => void;
  isSelected: boolean;
  isCompact: boolean;
  variant: TabsVariant;
};

function TabsItem<T extends string | number>({
  onSelect,
  isCompact,
  isSelected,
  item,
  variant,
}: TabsItemProps<T>) {
  const handlePress = () => onSelect(item.key);
  const resolvedStyles = getResolvedStyles(variant);

  return (
    <TouchableWithoutFeedback
      testID={TABS_ITEM_TEST_IDS.containerBtn(item.value, isSelected)}
      onPress={handlePress}
    >
      <ThemedView
        style={[
          styles.container,
          resolvedStyles.container,
          {
            ...(isSelected && resolvedStyles.containerSelected),
            ...(isCompact ? compactStyles.container : nonCompactStyles.container),
          },
        ]}
      >
        <ThemedText
          variant="header-sm"
          style={{
            ...styles.text,
            ...resolvedStyles.text,
            ...(isSelected && resolvedStyles.textActive),
          }}
          testID={TABS_ITEM_TEST_IDS.text(item.value)}
        >
          {item.value}
        </ThemedText>
      </ThemedView>
    </TouchableWithoutFeedback>
  );
}

const getResolvedStyles = (variant: TabsVariant) => {
  const variantMap = {
    default: defaultStyles,
    inverted: invertedStyles,
    'default-separated': defaultSeparatedStyles,
  } satisfies Record<TabsVariant, any>;

  if (!variantMap[variant]) throw new Error(`Invalid variant: ${variant}`);

  return variantMap[variant];
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 100,
  },
  text: {
    textTransform: 'capitalize',
  },
});

const compactStyles = StyleSheet.create({
  container: {
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
});

const nonCompactStyles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 16,
  },
});

const defaultStyles = StyleSheet.create({
  container: {
    backgroundColor: 'transparent',
  },
  containerSelected: {
    backgroundColor: colors.v2.white,
  },
  text: {
    color: colors.v2.white,
  },
  textActive: {
    color: colors.v2.black,
  },
});

const invertedStyles = StyleSheet.create({
  container: {
    backgroundColor: 'transparent',
  },
  containerSelected: {
    backgroundColor: colors.white,
  },
  text: {
    color: colors.white,
  },
  textActive: {
    color: colors.black,
  },
});

const defaultSeparatedStyles = StyleSheet.create({
  container: {
    backgroundColor: colors.v2.grayDark,
  },
  containerSelected: {
    backgroundColor: colors.black,
  },
  text: {
    color: colors.black,
  },
  textActive: {
    color: colors.white,
  },
});

export default TabsItem;
