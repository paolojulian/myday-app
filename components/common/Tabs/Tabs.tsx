import TabsItem, { TabItem } from '@/components/common/Tabs/TabsItem';
import ThemedView from '@/components/common/ThemedView';
import { colors } from '@/constants/Colors';
import { StyleSheet } from 'react-native';

export type TabsVariant = 'default' | 'inverted' | 'default-separated';

type TabsProps<T extends string | number> = {
  onSelect: (item: T) => void;
  selectedItem: T;
  items: TabItem<T>[];
  variant?: TabsVariant;
  isCompact?: boolean;
};

export default function Tabs<T extends string | number>({
  onSelect,
  selectedItem,
  items,
  variant = 'default',
  isCompact = false,
}: TabsProps<T>) {
  const resolvedStyles = getResolvedStyles(variant);

  return (
    <ThemedView
      style={[styles.container, resolvedStyles.container, { ...(isCompact ? styles.compact : {}) }]}
    >
      {items.map(item => (
        <TabsItem<T>
          key={item.key}
          onSelect={onSelect}
          isSelected={selectedItem === item.key}
          isCompact={isCompact}
          item={item}
          variant={variant}
        />
      ))}
    </ThemedView>
  );
}

const getResolvedStyles = (variant: TabsVariant) => {
  const variantMap: Record<TabsVariant, any> = {
    default: defaultStyles,
    inverted: invertedStyles,
    'default-separated': defaultSeparatedStyles,
  };

  if (!variantMap[variant]) throw new Error(`Invalid variant: ${variant}`);

  return variantMap[variant];
};

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    borderRadius: 100,
  },
  compact: {
    gap: 8,
  },
});

const defaultStyles = StyleSheet.create({
  container: {
    backgroundColor: colors.v2.grayDark,
  },
});

const invertedStyles = StyleSheet.create({
  container: {
    backgroundColor: colors.black,
  },
});

const defaultSeparatedStyles = StyleSheet.create({
  container: {
    backgroundColor: 'transparent',
  },
});
