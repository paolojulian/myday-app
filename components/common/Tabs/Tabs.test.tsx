import Tabs from '@/components/common/Tabs/Tabs';
import { render } from '@testing-library/react-native';

describe('TESTING Tabs', () => {
  const items: string[] = ['item1', 'item2', 'item3'];
  describe('WHEN rendering', () => {
    it('should match snapshot', () => {
      const component = render(
        <Tabs<string> items={items} onSelect={jest.fn} selectedItem={'item1'} />,
      );
      expect(component).toMatchSnapshot();
    });
    it('should render correctly', () => {
      const component = render(
        <Tabs<string> items={items} onSelect={jest.fn} selectedItem={'item1'} />,
      );
      expect(component).toBeTruthy();
    });
  });

  describe('WHEN selecting an item', () => {
    it('should call onSelect with the selected item', () => {});
  });
});
