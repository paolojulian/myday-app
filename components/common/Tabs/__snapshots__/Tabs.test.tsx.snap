// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TESTING Tabs WHEN rendering should match snapshot 1`] = `
<View
  style={
    [
      {
        "borderRadius": 100,
        "display": "flex",
        "flexDirection": "row",
      },
      {
        "backgroundColor": "#404040",
      },
      {},
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 100,
          "justifyContent": "center",
        },
        {
          "backgroundColor": "transparent",
        },
        {
          "backgroundColor": "#fafafa",
          "flex": 1,
          "padding": 16,
        },
      ]
    }
    testID="tabs-item-container-btn-item1-true"
  >
    <Text
      style={
        [
          {
            "fontFamily": "RalewayBold",
            "fontSize": 16,
            "fontWeight": 700,
          },
          {
            "color": "#fafafa",
          },
          {
            "fontWeight": 400,
          },
          {
            "color": "#171717",
            "textTransform": "capitalize",
          },
        ]
      }
      testID="tabs-item-text-item1"
    >
      item1
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 100,
          "justifyContent": "center",
        },
        {
          "backgroundColor": "transparent",
        },
        {
          "flex": 1,
          "padding": 16,
        },
      ]
    }
    testID="tabs-item-container-btn-item2-false"
  >
    <Text
      style={
        [
          {
            "fontFamily": "RalewayBold",
            "fontSize": 16,
            "fontWeight": 700,
          },
          {
            "color": "#fafafa",
          },
          {
            "fontWeight": 400,
          },
          {
            "color": "#fafafa",
            "textTransform": "capitalize",
          },
        ]
      }
      testID="tabs-item-text-item2"
    >
      item2
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 100,
          "justifyContent": "center",
        },
        {
          "backgroundColor": "transparent",
        },
        {
          "flex": 1,
          "padding": 16,
        },
      ]
    }
    testID="tabs-item-container-btn-item3-false"
  >
    <Text
      style={
        [
          {
            "fontFamily": "RalewayBold",
            "fontSize": 16,
            "fontWeight": 700,
          },
          {
            "color": "#fafafa",
          },
          {
            "fontWeight": 400,
          },
          {
            "color": "#fafafa",
            "textTransform": "capitalize",
          },
        ]
      }
      testID="tabs-item-text-item3"
    >
      item3
    </Text>
  </View>
</View>
`;
